### PR TITLE
[efficiency-improver] perf: eliminate redundant ContainsKey in IPC deserialization hot path

### PR DESCRIPTION
**Pull request by nohwnd**
## Goal and Rationale

Eliminate redundant dictionary hash lookups in the IPC deserialization hot path (`JsoniteConvert`). Both `DeserializeTestCase` and `DeserializeTestResult` used `ContainsKey` to detect the wire-format version ("flat" vs "nested"), then immediately repeated the same key lookup with `TryGetValue` — hashing the same key twice per test case/result received over IPC.

**Focus area:** Network & I/O Efficiency (IPC receive path)

---

## Approach

1. **`DeserializeTestCase`** — replaced `dict.ContainsKey("FullyQualifiedName")` with `dict.TryGetValue("FullyQualifiedName", out var fqnFlat)`, reusing `fqnFlat` directly instead of a second `TryGetValue` call on the same key.

2. **`DeserializeTestResult`** — replaced `dict.ContainsKey("Outcome")` with `dict.TryGetValue("Outcome", out var ocFlat)`, and removed the now-redundant inner `if (dict.TryGetValue("Outcome", ...))` guard.

Both changes are semantics-preserving: the `flat` boolean has the same value, and the extracted value is used identically.

---

## Energy Efficiency Evidence

**Proxy metric used: CPU instruction count / hash operations** → fewer redundant dictionary hash computations = fewer CPU cycles per test case/result received.

| Location | Before | After |
|---|---|---|
| `DeserializeTestCase` | 2 hash lookups for `"FullyQualifiedName"` per deserialized test case | 1 hash lookup |
| `DeserializeTestResult` | 2 hash lookups for `"Outcome"` per deserialized test result | 1 hash lookup |

For a **10 000-test run** (discovery + execution):
- **~20 000 redundant dictionary hash lookups eliminated** — 10K in discovery (test cases) + 10K in execution (test results)
- These are called in the IPC receive loop — every test case discovered and every test result received takes this path

---

## Green Software Foundation Context

- **Hardware Efficiency**: fewer CPU instructions per test case received over IPC reduces overall CPU energy per `dotnet test` invocation. At ecosystem scale (millions of CI runs daily), even small per-invocation reductions accumulate.
- **SCI (Software Carbon Intensity)**: reduces CPU work per functional unit (test run), lowering the energy term in the SCI equation.

---

## Trade-offs

None. This is a pure refactoring that preserves identical behaviour. The change is a 4-line diff with no added complexity.

---

## Reproducibility

```bash
./build.sh
./test.sh -p CommunicationUtilities
```

## Test Status

✅ Build succeeded (0 errors, 0 new warnings from this change)
✅ All 630 `Microsoft.TestPlatform.CommunicationUtilities.UnitTests` tests passed (net11.0|x64)

> Generated by [Efficiency Improver](https://github.com/microsoft/vstest/actions/runs/28188098090) · 1.1K AIC · ⌖ 26.4 AIC · ⊞ 45.6K · [◷](https://github.com/search?q=repo%3Amicrosoft%2Fvstest+%22gh-aw-workflow-id%3A+efficiency-improver%22&type=pullrequests)

### CODE REVIEW

File: src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/JsoniteConvert.cs [hunk 1]
```diff
@@ -288,10 +288,10 @@ private static object SerializeTestExecutionContext(TestExecutionContext c, Hash
         var tr = new TestResult(tc);
         if (dict.TryGetValue("Attachments", out var ao) && ao is IList al) foreach (var i in al) if (i != null && DeserializeAttachmentSet(i) is AttachmentSet a) tr.Attachments.Add(a);
         if (dict.TryGetValue("Messages", out var mo) && mo is IList ml) foreach (var i in ml) if (i != null && ConvertTo(i, typeof(TestResultMessage)) is TestResultMessage m) tr.Messages.Add(m);
-        bool flat = dict.ContainsKey("Outcome");
+        bool flat = dict.TryGetValue("Outcome", out var ocFlat);
         if (flat)
         {
-            if (dict.TryGetValue("Outcome", out var oc)) tr.Outcome = (TestOutcome)Convert.ToInt32(oc, CultureInfo.InvariantCulture);
+            tr.Outcome = (TestOutcome)Convert.ToInt32(ocFlat, CultureInfo.InvariantCulture);
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/JsoniteConvert.cs:294]: [Null Safety & Boundary Validation / Algorithmic Correctness]

Removing the inner guard is correct — `dict.TryGetValue("Outcome", ...)` inside `if (flat)` was always `true` (key presence is guaranteed by the outer check), so the guard was purely redundant. The new code is semantically identical across all cases:

| "Outcome" key | value | Old result | New result |
|---|---|---|---|
| absent | — | `flat=false`, block skipped | same |
| present | `1` | `TestOutcome.Passed` | same |
| present | `null` | `Convert.ToInt32(null)=0` → `TestOutcome.None` | same |

The null→0 coercion is a `Convert.ToInt32(object, IFormatProvider)` contract detail (documented: returns 0 when value is null). Old `oc` and new `ocFlat` were always on the same path. No behavioral change here.

nit: contrast with `DeserializeTestCase` line 252 where an explicit `if (fqnFlat != null)` guard is kept — the asymmetry is intentional (skipping a `string` assignment vs. defaulting an `enum` to 0), but a brief inline comment would make the null-tolerance conscious for future readers:

```suggestion
            tr.Outcome = (TestOutcome)Convert.ToInt32(ocFlat, CultureInfo.InvariantCulture); // null → 0 (TestOutcome.None) by Convert contract
```

**Review by nohwnd — COMMENTED** [review]
## Review: `JsoniteConvert` IPC deserialization optimization

### Dimension sweep

| Dimension | Status | Notes |
|---|---|---|
| IPC Transport & Protocol Stability | ✅ pass | Deserialization-only; no wire format touched |
| Backward Compatibility & Rollback Safety | ✅ pass | Semantics preserved across all cases |
| Null Safety & Boundary Validation | ⚠️ minor | See inline — implicit null-to-0 coercion, same as before |
| Algorithmic Correctness | ✅ pass | `TryGetValue` bool ≡ `ContainsKey` bool; inner guard removal valid |
| Cross-TFM & Framework Resolution | ✅ pass | Correctly scoped to `#if !NETCOREAPP` (line 4) |
| Performance | i️ info | Optimization is real; one fewer hash lookup per call |
| Acceptance Test Coverage | ✅ pass | See below |

### Correctness

Both changes are semantics-preserving. `TryGetValue` returns `true` iff the key exists — identical to `ContainsKey` for any conforming `IDictionary` implementation. The captured `fqnFlat`/`ocFlat` values are used exactly as the old `fqn`/`oc` were:

- `DeserializeTestCase`: `if (fqnFlat != null)` guard preserved. ✓
- `DeserializeTestResult`: inner guard dropped correctly — it was always `true` inside `if (flat)`, making it a no-op. See inline comment for the null-edge-case trace.

### Test coverage

Better than it looks. `SerializationGoldenTests.cs` targets `net481` (in addition to `net11.0`), so `TestResultRoundTripV2ShouldPreserveKeyProperties` and `TestCaseRoundTripV2ShouldPreserveKeyProperties` exercise the exact Jsonite `DeserializeTestCase`/`DeserializeTestResult` flat paths changed here — and both assert field values including `Outcome` and `FullyQualifiedName`. No new test is strictly needed for the happy path.

The null-valued-key edge case (`"Outcome": null` on the wire) isn't tested, but the behavior is unchanged from before, and in practice the IPC protocol never emits a null Outcome.

### Performance claim

Valid. One fewer `Dictionary<string, object>` hash lookup per TestCase/TestResult received over IPC. The magnitude is nanoseconds per call — unmeasured and unmeasurable at the test-run level — but the claim is directionally correct and the change introduces no new allocations.

One nit inline — otherwise looks good :)

> 🧠 *Reviewed by [Expert Code Reviewer 🧠](https://github.com/microsoft/vstest/actions/runs/28189006385)*

**Review by nohwnd — COMMENTED** [review]
## Review: `JsoniteConvert` IPC deserialization optimization (re-run)

### Dimension sweep

| Dimension | Status | Notes |
|---|---|---|
| IPC Transport & Protocol Stability | ✅ pass | Deserialization-only; no wire format change |
| Backward Compatibility & Rollback Safety | ✅ pass | Semantics preserved across all key/value combinations |
| Null Safety & Boundary Validation | i️ nit | `Convert.ToInt32(null)=0` is documented contract; prior inline nit (line 294) still unresolved |
| Algorithmic Correctness | ✅ pass | `TryGetValue` bool ≡ `ContainsKey` bool; inner guard removal valid |
| Cross-TFM & Framework Resolution | ✅ pass | Entire file is `#if !NETCOREAPP` — change is correctly scoped |
| Performance | i️ info | One fewer hash lookup per TestCase/TestResult deserialized; no new allocations |

### Correctness

Both changes are semantics-preserving under full inspection of the surrounding context:

**`DeserializeTestCase` (line 249):** `TryGetValue` returns `true` iff the key exists — identical to `ContainsKey`. The captured `fqnFlat` is used identically to the prior `fqn`. The `if (fqnFlat != null)` guard is preserved. When `flat=false`, `fqnFlat` is `null` but is never read. ✅

**`DeserializeTestResult` (line 291):** Same `TryGetValue`/`ContainsKey` equivalence. The removed inner guard `if (dict.TryGetValue("Outcome", out var oc))` was always `true` inside `if (flat)` — key presence is guaranteed by the outer check. `Convert.ToInt32(null, InvariantCulture) = 0 = TestOutcome.None` matches prior behavior when `oc` was null. ✅

The `flat` / `!flat` dispatch at lines 267 and 310 is unaffected.

### Prior inline comment

A nit was posted in the previous review cycle on line 294 suggesting a comment to document the `null → 0 (TestOutcome.None)` coercion. It remains unresolved and is not a blocking concern.

> 🧠 *Reviewed by [Expert Code Reviewer 🧠](https://github.com/microsoft/vstest/actions/runs/28236768318)*

> 🧠 *Reviewed by [Expert Code Reviewer 🧠](https://github.com/microsoft/vstest/actions/runs/28236768318)*

**Review by azat-msft — APPROVED** [review]
